kernel: pipe: read_avail / write_avail syscalls

cfriedt opened the pull request:
These helper functions may clarify how much data is available to read
from or write to the circular buffer of a k_pipe.

Fixes #25036

andyross commented on include/kernel.h:
```diff
@@ -4430,6 +4430,56 @@ __syscall int k_pipe_get(struct k_pipe *pipe, void *data,
+	if (pipe->read_index == pipe->write_index) {
+		return pipe->bytes_used;
+	} else if (pipe->read_index < pipe->write_index) {
+		return pipe->write_index - pipe->read_index;
```
Isn't the pipe a ring?  I think this subtraction needs to be modulus the size, no?

cfriedt replied:
That's the "invisible else" condition, but we avoid the modulus by changing the order of the operands.

cfriedt replied:
It is a ring. If read < write, then it's the straight difference. Otherwise, we would use the modulus operator, but we can get away from a potentially expensive division in hw by subtracting from total size.
The read == write is a special case.
I wish there was a way to do this without branching (i.e. stalling the pipeline). Bit-twiddling hacks can be pretty awesome.

andyross reviewed (commented):
Heh, I suspected when I saw that enhancement request go by that there'd be a patch right behind it.

One quibble that doesn't match my understanding of how the buffer works, but I might be wrong and will trust your answer.

tyrantlizard77 reviewed (changes requested):
This PR needs:

1) verification functions for the system calls
2) test cases that prove that these new APIs work, with the test cases running in user mode so that there is code coverage of both the APIs and the verification functions

https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html

andyross reviewed (approved):
Got it.  Looks great.

andrewboie commented:
Code looks good BTW, just need tests for any new kernel APIs...can add them to one of the suites in tests/kernel/pipe (no I don't know why there are two of them, pick one I suppose)

pfalcon reviewed (approved):
Thanks for submitting. To make it explicit, this was split off from https://github.com/zephyrproject-rtos/zephyr/pull/24813 (which relies on it) to streamline the review process.

cfriedt commented:
> This PR needs:
> 
> 1. verification functions for the system calls
> 2. test cases that prove that these new APIs work, with the test cases running in user mode so that there is code coverage of both the APIs and the verification functions
> 
> https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html

I saw that there were a few inline syscalls, so thought I might be able to get away without the verify function, but I will do it shortly.

@andrewboie - do you recommend also taking the spinlock in that case?

They will not be inline if verify functions are used and will be inside kernel/pipes.c instead.

andrewboie commented:
> I saw that there were a few inline syscalls, so thought I might be able to get away without the verify function, but I will do it shortly.

Inline syscalls still have non-inline verification functions. If the verification functions don't exist, you get a fatal "unimplemented system call" error if you try to invoke it from user mode.

> @andrewboie - do you recommend also taking the spinlock in that case?

Oh, I didn't notice this earlier but yeah, you could get incorrect or outright goofy results if these APIs are called at just the wrong time, like if another CPU is modifying the same data...I think using the lock is safest.

pfalcon commented:
> if these APIs are called at just the wrong time

We could also start with introducing these APIs as kernel-only, and wait for actual usecases for using them from userspace. The current usecase is implementing higher-level synchro primitives in kernel space (so locking would also be handled by "dependency injection" (i.e. at a higher level)). Just thinking aloud...

andrewboie commented:
> > if these APIs are called at just the wrong time
> 
> We could also start with introducing these APIs as kernel-only, and wait for actual usecases for using them from userspace. The current usecase is implementing higher-level synchro primitives in kernel space (so locking would also be handled by "dependency injection" (i.e. at a higher level)). Just thinking aloud...

The potential concurrency issues we are seeing has *nothing* to do with user mode. Two supervisor threads can have the same problems.

cfriedt commented on tests/kernel/pipe/pipe_api/src/test_pipe_avail.c:
```diff
@@ -0,0 +1,137 @@
+static struct k_pipe bufferless;
+
+/*
+ *  r	 w
```
Sorry - spacing is a bit off in GH. It's correct with monospaced font.

cfriedt commented:
Should be good to go now. Apologies, but the spacing is off a little in the test_pipe_avail.c; the 'r' and 'w' line up properly with a monospace font.

andrewboie commented on kernel/pipes.c:
```diff
@@ -805,3 +805,77 @@ void k_pipe_block_put(struct k_pipe *pipe, struct k_mem_block *block,
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_pipe_read_avail(struct k_pipe *pipe)
```
should be size_t return value

cfriedt replied:
Ack! Sorry

andrewboie commented on kernel/pipes.c:
```diff
@@ -805,3 +805,77 @@ void k_pipe_block_put(struct k_pipe *pipe, struct k_mem_block *block,
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_pipe_write_avail(struct k_pipe *pipe)
```
should be size_t

tyrantlizard77 reviewed (commented):
looks great just a nit on return value types

pfalcon commented:
> The potential concurrency issues we are seeing has nothing to do with user mode. Two supervisor threads can have the same problems.

Yes, and I mean that these functions could stay kernel-only with a requirement that synchronization should be handled by the caller. I may imagine, that would work for the current usecase of https://github.com/zephyrproject-rtos/zephyr/pull/24813.

Of course, if we shoot straight for the userspace support of these APIs, we don't have a choice but handle locking issues right now.

andrewboie commented:
> > The potential concurrency issues we are seeing has nothing to do with user mode. Two supervisor threads can have the same problems.
> 
> Yes, and I mean that these functions could stay kernel-only with a requirement that synchronization should be handled by the caller. I may imagine, that would work for the current usecase of #24813.

NACK

cfriedt committed:
kernel: pipe: read_avail / write_avail syscalls

These helper functions may clarify how much data is available to read
from or write to the circular buffer of a k_pipe.

Fixes #25036

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
Looks like there is a quite unrelated failure in the kinetis frdm board build.

I don't think it should block.

Anyone have privs to merge anyway?
```
[117/138] Building C object zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj
FAILED: zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj 
ccache /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBOARD_FLASH_SIZE=CONFIG_FLASH_SIZE -DBUILD_VERSION=zephyr-v2.2.0-2427-gbf9c4ea1568e -DCPU_MK64FN1M0VLL12 -DKERNEL -DNDEBUG -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I../../../../../../include -Izephyr/include/generated -I../../../../../../soc/arm/nxp_kinetis/k6x -I../../../../../../lib/fnmatch/. -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/cmsis/CMSIS/Core/Include -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/nxp/mcux/devices/MK64F12 -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/nxp/mcux/drivers/kinetis/. -isystem ../../../../../../lib/libc/minimal/include -isystem /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include -isystem /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include-fixed -Os -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/frdm_k64f/tests/boards/board_shell/boards.shell.shell/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -mcpu=cortex-m4 -mthumb -mabi=aapcs -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-address-of-packed-member -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/boards/board_shell=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj -MF zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj.d -o zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/adc/adc_shell.c
In file included from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/adc/adc_shell.c:7:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/adc/adc_shell.c:477:22: error: expected expression before ',' token
  477 |  ADC_SHELL_COMMAND(0),
      |                      ^
../../../../../../include/shell/shell.h:219:3: note: in definition of macro 'SHELL_STATIC_SUBCMD_SET_CREATE'
  219 |   __VA_ARGS__      \
      |   ^~~~~~~~~~~
```

nashif commented on tests/kernel/pipe/pipe_api/src/test_pipe_avail.c:
```diff
@@ -0,0 +1,137 @@
+ * w_avail = N - (w - r) = 5
+ * would overwrite: e f g h
+ */
+void test_pipe_avail_r_lt_w(void)
```
no enforced or a common practice yet, but it would be great to have doxygen documentation for all tests (those that start with test_). We are going to use this for generating test plans and test documentation collateral. Basically a brief of what is being tested and short description of the test workflow. 

Thanks.

carlescufi replied:
@cfriedt could you please add the doxygen for those tests? Should be relatively simple, a matter of just documenting the purpose in the context of a Doxygen comment.

cfriedt replied:
Will do.

cfriedt commented:
Can someone add the #v2.3.0 milestone to this ticket? I'd like to get it merged in by then.

carlescufi commented:
> Can someone add the #v2.3.0 milestone to this ticket? I'd like to get it merged in by then.

I have already

cfriedt committed:
tests: kernel: pipe: tests for read_avail write_avail

Tests for issue #25036

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>
